src/sync/impl/sync_metadata.cpp

```diff
@@ -37,52 +37,95 @@ static const char * const c_sync_marked_for_removal = "marked_for_removal";
 
+namespace {
 
-SyncMetadataManager::SyncMetadataManager(std::string path,
-                                         bool should_encrypt,
-                                         util::Optional<std::vector<char>> encryption_key)
+Property make_nullable_string_property(const char* name)
 {
-    std::lock_guard<std::mutex> lock(m_metadata_lock);
+    Property p = {name, PropertyType::String};
+    p.is_nullable = true;
+    return p;
+}
 
-    auto make_nullable_string_property = [](const char *name) -> Property {
-        Property p = {name, PropertyType::String};
-        p.is_nullable = true;
-        return p;
-    };
+Property make_primary_key_property(const char* name)
+{
+    Property p = {name, PropertyType::String};
+    p.is_indexed = true;
+    p.is_primary = true;
+    return p;
+}
 
-    auto make_primary_key_property = [](const char *name) -> Property {
-        Property p = {name, PropertyType::String};
-        p.is_indexed = true;
-        p.is_primary = true;
-        return p;
-    };
+ObjectSchema make_v0_fileActionMetadata_schema()
+{
+    return {c_sync_fileActionMetadata, {
+        make_primary_key_property(c_sync_original_name),
+        {c_sync_action, PropertyType::Int},
+        make_nullable_string_property(c_sync_new_name),
+        {c_sync_url, PropertyType::String},
+        {c_sync_identity, PropertyType::String},
+    }};
+}
 
-    Realm::Config config;
-    config.path = std::move(path);
-    config.schema = Schema{
+Schema make_v0_schema()
+{
+    return Schema{
         {c_sync_userMetadata, {
             make_primary_key_property(c_sync_identity),
             {c_sync_marked_for_removal, PropertyType::Bool},
             make_nullable_string_property(c_sync_auth_server_url),
             make_nullable_string_property(c_sync_user_token),
         }},
-        {c_sync_fileActionMetadata, {
-            make_primary_key_property(c_sync_original_name),
-            {c_sync_action, PropertyType::Int},
-            make_nullable_string_property(c_sync_new_name),
-            {c_sync_url, PropertyType::String},
-            {c_sync_identity, PropertyType::String},
+        make_v0_fileActionMetadata_schema(),
+    };
+}
+
+Schema make_v1_schema()
+{
+    return Schema{
+        {c_sync_userMetadata, {
+            make_primary_key_property(c_sync_identity),
+            {c_sync_marked_for_removal, PropertyType::Bool},
+            make_nullable_string_property(c_sync_auth_server_url),
+            make_nullable_string_property(c_sync_user_token),
+            {c_sync_user_is_admin, PropertyType::Bool},
         }},
+        make_v0_fileActionMetadata_schema(),
     };
+}
+
+}
+
+#pragma mark - Sync metadata manager
+
+SyncMetadataManager::SyncMetadataManager(std::string path,
+                                         bool should_encrypt,
+                                         util::Optional<std::vector<char>> encryption_key,
+                                         int schema_version)
```

Comment:
It looks like the support for multiple schema versions is only needed for sake of testing the upgrade path. IMO the logic to create Realms with an older schema version should live in the unit tests, or we should test against prepackaged Realms created using the old code.